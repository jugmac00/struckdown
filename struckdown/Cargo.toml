[package]
name = "struckdown"
version = "0.1.0"
authors = ["Armin Ronacher <armin.ronacher@active-4.com>"]
edition = "2018"
license = "Apache-2.0"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[features]
default = ["external-processor"]
external-processor = ["tokio", "subprocess"]

[dependencies]
pulldown-cmark = "0.8.0"
serde = { version = "1.0.118", features = ["derive"] }
regex = "1.4.2"
lazy_static = "1.4.0"
v_htmlescape = "0.11.0"
itertools = "0.9.0"
serde_yaml = "0.8.14"
slug = "0.1.4"
serde_json = { version = "1.0.60", features = ["preserve_order"] }
subprocess = { version = "0.2.6", optional = true }
tokio = { version = "0.3.6", features = ["rt", "process", "macros", "io-util"], optional=true }

[dev-dependencies]
insta = { version = "1.3.0", features = ["glob"] }
