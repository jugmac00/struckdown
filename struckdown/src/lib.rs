//! struckdown is a text processing system based on markdown/commonmark with
//! structured extensions inspired by reStructuredText.  It parses commonmark
//! with structural extensions (roles and directives) into an event stream.
//!
//! This event stream can be processed before rendering.
pub mod event;
pub mod html;
pub mod parser;
pub mod pipeline;
pub mod processors;

// internal only for now
mod plain;

/// Gives access to [`serde_json`] value functionality.
///
/// The [`Value`](crate::value::Value) type is used to represent arbitrary data in a few instances.
/// A notable example is emitted [`front_matter`](crate::event::DocumentStartEvent::front_matter)
/// or other [`MetaData`](crate::event::MetaDataEvent).
pub mod value {
    pub use serde_json::value::*;

    /// A re-export of the [`serde_json::json`] macro.
    pub use serde_json::json as value;

    pub use serde_json::{from_value, to_value};
}
