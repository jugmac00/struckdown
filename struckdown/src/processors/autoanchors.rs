use slug::slugify;
use std::collections::VecDeque;

use crate::event::{AnnotatedEvent, Event, StartTagEvent, Tag};

/// Adds anchors to all headers if missing.
pub struct AutoAnchors<'data, I> {
    source: I,
    buffer: VecDeque<AnnotatedEvent<'data>>,
    min_level: usize,
}

impl<'data, I: Iterator<Item = AnnotatedEvent<'data>>> AutoAnchors<'data, I> {
    /// Creates a new auto anchor stream processor.
    ///
    /// This sets the default minimum level to 6 which means that all headers
    /// will get an ID added.  This minimum level can be reconfigured by using
    /// the [`AutoAnchors::set_min_level`] method.
    pub fn new(iterator: I) -> Self {
        Self {
            source: iterator,
            buffer: VecDeque::new(),
            min_level: 6,
        }
    }

    /// Overrides the minimum level that should get IDs
    pub fn set_min_level(&mut self, min_level: usize) {
        self.min_level = min_level;
    }
}

impl<'data, I: Iterator<Item = AnnotatedEvent<'data>>> Iterator for AutoAnchors<'data, I> {
    type Item = AnnotatedEvent<'data>;

    fn next(&mut self) -> Option<Self::Item> {
        if let Some(annotated_event) = self.buffer.pop_front() {
            return Some(annotated_event);
        }

        self.source.next().map(|mut annotated_event| {
            let (header_level, attrs) = match annotated_event.event {
                Event::StartTag(StartTagEvent {
                    tag: Tag::Heading1,
                    ref mut attrs,
                }) => (1, attrs),
                Event::StartTag(StartTagEvent {
                    tag: Tag::Heading2,
                    ref mut attrs,
                }) => (2, attrs),
                Event::StartTag(StartTagEvent {
                    tag: Tag::Heading3,
                    ref mut attrs,
                }) => (3, attrs),
                Event::StartTag(StartTagEvent {
                    tag: Tag::Heading4,
                    ref mut attrs,
                }) => (4, attrs),
                Event::StartTag(StartTagEvent {
                    tag: Tag::Heading5,
                    ref mut attrs,
                }) => (5, attrs),
                Event::StartTag(StartTagEvent {
                    tag: Tag::Heading6,
                    ref mut attrs,
                }) => (6, attrs),
                _ => return annotated_event,
            };

            if attrs.id.is_some() || header_level > self.min_level {
                return annotated_event;
            }

            let mut depth = 1;
            let mut raw_text = String::new();

            while let Some(next_annotated_event) = self.source.next() {
                match next_annotated_event.event {
                    Event::StartTag(..) => depth += 1,
                    Event::EndTag(..) => depth -= 1,
                    ref event => {
                        if let Some(text) = event.raw_text() {
                            raw_text.push_str(text.as_str());
                        }
                    }
                }
                if depth == 0 {
                    break;
                }
            }

            attrs.id = Some(slugify(raw_text).into());

            annotated_event
        })
    }
}
