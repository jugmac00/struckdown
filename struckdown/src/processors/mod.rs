//! Stream processors for struckdown.
//!
//! This module provides stream processors that manipulate a struckdown event
//! stream to enhance it.  For instance a stream processor can automatically
//! add anchors to headers if they did not already get a header set by other
//! means.
mod autoanchors;

pub use self::autoanchors::{AutoAnchors, AutoAnchorsOptions};
