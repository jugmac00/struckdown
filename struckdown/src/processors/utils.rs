#[macro_export]
macro_rules! implement_processor {
    ($type:ty, $iter:tt) => {
        impl $crate::processors::Processor for $type {
            fn apply<'data>(
                self: Box<Self>,
                iter: Box<dyn Iterator<Item = $crate::event::AnnotatedEvent<'data>> + 'data>,
            ) -> Box<dyn Iterator<Item = $crate::event::AnnotatedEvent<'data>> + 'data> {
                Box::new($iter::new(iter, std::borrow::Cow::Owned(*self)))
            }

            fn apply_ref<'data, 'options: 'data>(
                &'options self,
                iter: Box<dyn Iterator<Item = $crate::event::AnnotatedEvent<'data>> + 'data>,
            ) -> Box<dyn Iterator<Item = $crate::event::AnnotatedEvent<'data>> + 'data> {
                Box::new($iter::new(iter, std::borrow::Cow::Borrowed(self)))
            }
        }
    };
}
