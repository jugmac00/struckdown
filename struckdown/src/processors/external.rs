use serde::{Deserialize, Serialize};
use std::borrow::Cow;
use std::path::PathBuf;
use std::process::Stdio;
use tokio::runtime::Runtime;

use tokio::io::BufReader;
use tokio::io::{AsyncBufReadExt, AsyncWriteExt};
use tokio::process::{ChildStdin, ChildStdout, Command};

use crate::event::AnnotatedEvent;
use crate::processors::Processor;

/// Passes a JSON serialized stream through an external program.
///
/// When applied this wraps the stream in a [`ExternalIter`].
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct External {
    /// The executable to run.
    pub cmd: PathBuf,
    /// The arguments to pass to the external command.
    #[serde(default)]
    pub args: Vec<String>,
}

impl Processor for External {
    fn apply<'data>(
        self: Box<Self>,
        iter: Box<dyn Iterator<Item = AnnotatedEvent<'data>> + 'data>,
    ) -> Box<dyn Iterator<Item = AnnotatedEvent<'data>> + 'data> {
        Box::new(ExternalIter::new(iter, Cow::Owned(*self)))
    }

    fn apply_ref<'data, 'options: 'data>(
        &'options self,
        iter: Box<dyn Iterator<Item = AnnotatedEvent<'data>> + 'data>,
    ) -> Box<dyn Iterator<Item = AnnotatedEvent<'data>> + 'data> {
        Box::new(ExternalIter::new(iter, Cow::Borrowed(self)))
    }
}

/// The iterator implementing [`External`].
pub struct ExternalIter<'data, 'options, I: Iterator<Item = AnnotatedEvent<'data>>> {
    source: I,
    spawned: bool,
    stdin: Option<ChildStdin>,
    stdout: Option<BufReader<ChildStdout>>,
    buffered_event: Option<Vec<u8>>,
    options: Cow<'options, External>,
    rt: Option<Runtime>,
}

impl<'data, 'options, I: Iterator<Item = AnnotatedEvent<'data>>> ExternalIter<'data, 'options, I> {
    pub fn new(iterator: I, options: Cow<'options, External>) -> Self {
        let rt = tokio::runtime::Builder::new_current_thread()
            .enable_all()
            .build()
            .unwrap();
        Self {
            source: iterator,
            spawned: false,
            stdin: None,
            stdout: None,
            buffered_event: None,
            options,
            rt: Some(rt),
        }
    }
}

impl<'data, 'options, I: Iterator<Item = AnnotatedEvent<'data>>> Iterator
    for ExternalIter<'data, 'options, I>
{
    type Item = AnnotatedEvent<'data>;

    fn next(&mut self) -> Option<Self::Item> {
        // TODO: error handling
        // This entire code is a massive hack.  There must be a better way to do this but
        // i'm too lazy to figure this out right now for a small experiment.

        let rt = self.rt.take().unwrap();
        if !self.spawned {
            rt.block_on(async {
                let mut process = Command::new(&self.options.cmd)
                    .args(&self.options.args)
                    .stdin(Stdio::piped())
                    .stdout(Stdio::piped())
                    .spawn()
                    .unwrap();
                self.stdin = process.stdin.take();
                self.stdout = process.stdout.take().map(BufReader::new);
                self.spawned = true;
            });
        }

        let mut done = false;
        while !done {
            if self.buffered_event.is_none() {
                if let Some(event) = self.source.next() {
                    let mut serialized_event = serde_json::to_vec(&event).unwrap();
                    serialized_event.push(b'\n');
                    self.buffered_event = Some(serialized_event);
                } else {
                    // close stdin if we're done writing.
                    self.stdin.take();
                }
            }

            let mut stdin = self.stdin.take();
            let mut stdout = self.stdout.take().unwrap();
            let mut rv = None;

            rt.block_on(async {
                let mut line = String::new();

                let should_write = self.buffered_event.is_some();
                let write = async {
                    if let (Some(ref mut stdin), Some(ref buffered_event)) = (&mut stdin, &self.buffered_event) {
                        stdin.write(buffered_event).await.unwrap();
                    }
                };

                tokio::select! {
                    read = stdout.read_line(&mut line) => {
                        match read {
                            Ok(0) | Err(_) => done = true,
                            Ok(_) => {
                                let event: AnnotatedEvent<'_> = serde_json::from_str(&line).unwrap();
                                rv = Some(event);
                            }
                        }
                    }
                    _ = write, if should_write => {
                        self.buffered_event.take();
                    }
                }
            });

            self.stdin = stdin;
            self.stdout = Some(stdout);

            if let Some(rv) = rv {
                self.rt = Some(rt);
                return Some(rv);
            } else if done {
                // close stuff
                self.stdin.take();
                self.stdout.take();
                self.rt = Some(rt);
                return None;
            }
        }

        None
    }
}
