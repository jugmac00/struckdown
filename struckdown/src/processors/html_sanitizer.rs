use std::borrow::Cow;
use std::collections::{BTreeMap, HashSet};

use ammonia::Builder;
use lazy_static::lazy_static;
use regex::Regex;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::event::{AnnotatedEvent, Event};
use crate::processors::Processor;

lazy_static! {
    static ref MARKER_RE: Regex = Regex::new(r"\A\.\.\.([a-f0-9]{32})\.\.\.").unwrap();
}

/// Automatically add anchors to all headers when missing.
///
/// When applied this wraps the stream in a [`HtmlSanitizerIter`].
#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(default)]
pub struct HtmlSanitizer {
    /// Configures link rel
    link_rel: Option<String>,
    /// If configured the `class` attribute is permitted without filtering.
    allow_class: bool,
    /// If configured the `style` attribute is permitted without filtering.
    allow_style: bool,
    /// If set to `false` then comments are removed.
    allow_comments: bool,
}

impl Default for HtmlSanitizer {
    fn default() -> HtmlSanitizer {
        HtmlSanitizer {
            link_rel: Some("noopener noreferrer".into()),
            allow_class: false,
            allow_style: false,
            allow_comments: true,
        }
    }
}

impl Processor for HtmlSanitizer {
    fn apply<'data>(
        self: Box<Self>,
        iter: Box<dyn Iterator<Item = AnnotatedEvent<'data>> + 'data>,
    ) -> Box<dyn Iterator<Item = AnnotatedEvent<'data>> + 'data> {
        Box::new(HtmlSanitizerIter::new(iter, Cow::Owned(*self)))
    }

    fn apply_ref<'data, 'options: 'data>(
        &'options self,
        iter: Box<dyn Iterator<Item = AnnotatedEvent<'data>> + 'data>,
    ) -> Box<dyn Iterator<Item = AnnotatedEvent<'data>> + 'data> {
        Box::new(HtmlSanitizerIter::new(iter, Cow::Borrowed(self)))
    }
}

fn make_ammonia(options: &HtmlSanitizer) -> Builder {
    let mut ammonia = Builder::default();
    let mut clean_content_tags = HashSet::new();
    clean_content_tags.insert("script");

    if options.allow_class {
        ammonia.add_generic_attributes(&["class"]);
    }
    if options.allow_style {
        ammonia.add_generic_attributes(&["style"]);
        ammonia.add_tags(&["style"]);
    } else {
        clean_content_tags.insert("style");
    }
    ammonia.clean_content_tags(clean_content_tags);
    ammonia.link_rel(options.link_rel.as_deref());
    ammonia.strip_comments(!options.allow_comments);
    ammonia
}

/// The iterator implementing [`HtmlSanitizer`].
pub struct HtmlSanitizerIter<'data, 'options, I: Iterator<Item = AnnotatedEvent<'data>>> {
    source_iter: Option<I>,
    processed_iter: std::vec::IntoIter<AnnotatedEvent<'data>>,
    options: Cow<'options, HtmlSanitizer>,
}

impl<'data, 'options, I: Iterator<Item = AnnotatedEvent<'data>>>
    HtmlSanitizerIter<'data, 'options, I>
{
    pub fn new(iterator: I, options: Cow<'options, HtmlSanitizer>) -> Self {
        Self {
            source_iter: Some(iterator),
            processed_iter: Vec::new().into_iter(),
            options,
        }
    }
}

impl<'data, 'options, I: Iterator<Item = AnnotatedEvent<'data>>> Iterator
    for HtmlSanitizerIter<'data, 'options, I>
{
    type Item = AnnotatedEvent<'data>;

    fn next(&mut self) -> Option<Self::Item> {
        if let Some(source_iter) = self.source_iter.take() {
            let marker = format!("...{}...", Uuid::new_v4().to_simple());
            let mut buffer = source_iter.collect::<Vec<_>>();
            let mut html_buf = String::new();
            let ammonia = make_ammonia(&self.options);
            let mut segments = BTreeMap::new();
            let mut idx = 0;
            for annotated_event in &mut buffer {
                if let Event::RawHtml(ref mut raw_html) = annotated_event.event {
                    idx += 1;
                    let id = Uuid::new_v4();
                    html_buf.push_str(&format!("...{}...", id.to_simple()));
                    html_buf.push_str(raw_html.html.as_str());
                    html_buf.push_str(&marker);
                    segments.insert(id, idx);
                }
            }
            let cleaned = dbg!(ammonia.clean(&html_buf).to_string());

            let mut replacements = BTreeMap::new();
            for segment in cleaned.split(&marker) {
                if let Some(id) = MARKER_RE.captures(segment) {
                    if let Some(&idx) = segments.get(&Uuid::parse_str(&id[1]).unwrap()) {
                        replacements.insert(idx, &segment[id.get(0).unwrap().end()..]);
                    }
                }
            }

            let mut idx = 0;
            for annotated_event in &mut buffer {
                if let Event::RawHtml(ref mut raw_html) = annotated_event.event {
                    idx += 1;
                    if let Some(&new) = replacements.get(&idx) {
                        raw_html.html = new.to_string().into();
                        continue;
                    }
                    raw_html.html = "".into();
                }
            }

            self.processed_iter = buffer.into_iter();
        }
        self.processed_iter.next()
    }
}
