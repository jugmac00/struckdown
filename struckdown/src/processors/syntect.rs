use std::borrow::Cow;
use std::marker::PhantomData;
use std::path::PathBuf;

use serde::{Deserialize, Serialize};
use syntect::easy::HighlightLines;
use syntect::highlighting::{Theme, ThemeSet};
use syntect::html::{styled_line_to_highlighted_html, IncludeBackground};
use syntect::parsing::SyntaxSet;

use crate::event::{AnnotatedEvent, CodeBlockEvent, Event, RawHtmlEvent};
use crate::processors::Processor;

const DEFAULT_THEME: &str = "InspiredGitHub";

/// Implements syntax highlighting via [`syntect`].
///
/// When applied this wraps the stream in a [`SyntectIter`].
#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(default)]
pub struct Syntect {
    /// The name of the theme to use.  If both this and `theme_path` is not
    /// set then a default theme is loaded.  If `theme_path` is set then the
    /// theme in the `theme_path` folder is used.
    pub theme: Option<String>,
    /// When `theme` is not set, then the path to the `.tmTheme` file to load
    /// otherwise the folder to a collection of theme files.
    pub theme_path: Option<PathBuf>,
}

impl Default for Syntect {
    fn default() -> Syntect {
        Syntect {
            theme: None,
            theme_path: None,
        }
    }
}

impl Processor for Syntect {
    fn apply<'data>(
        self: Box<Self>,
        iter: Box<dyn Iterator<Item = AnnotatedEvent<'data>> + 'data>,
    ) -> Box<dyn Iterator<Item = AnnotatedEvent<'data>> + 'data> {
        Box::new(SyntectIter::new(iter, Cow::Owned(*self)))
    }

    fn apply_ref<'data, 'options: 'data>(
        &'options self,
        iter: Box<dyn Iterator<Item = AnnotatedEvent<'data>> + 'data>,
    ) -> Box<dyn Iterator<Item = AnnotatedEvent<'data>> + 'data> {
        Box::new(SyntectIter::new(iter, Cow::Borrowed(self)))
    }
}

/// The iterator implementing [`Syntect`].
pub struct SyntectIter<'data, 'options, I: Iterator<Item = AnnotatedEvent<'data>>> {
    source: I,
    syntax_set: SyntaxSet,
    theme: Theme,
    options: PhantomData<&'options Syntect>,
}

impl<'data, 'options, I: Iterator<Item = AnnotatedEvent<'data>>> SyntectIter<'data, 'options, I> {
    pub fn new(iterator: I, options: Cow<'options, Syntect>) -> Self {
        let theme = match (&options.theme, &options.theme_path) {
            (Some(theme), None) => {
                let mut theme_set = ThemeSet::load_defaults();
                match theme_set.themes.remove(theme) {
                    Some(theme) => theme,
                    None => theme_set.themes.remove(DEFAULT_THEME).unwrap(),
                }
            }
            (Some(theme), Some(path)) => {
                let mut theme_set =
                    ThemeSet::load_from_folder(path).expect("failed to initialized theme folder");
                match theme_set.themes.remove(theme) {
                    Some(theme) => theme,
                    None => theme_set.themes.remove(DEFAULT_THEME).unwrap(),
                }
            }
            (None, Some(ref path)) => {
                ThemeSet::get_theme(path).expect("failed to load theme by path")
            }
            (None, None) => {
                let mut theme_set = ThemeSet::load_defaults();
                theme_set.themes.remove(DEFAULT_THEME).unwrap()
            }
        };
        Self {
            source: iterator,
            syntax_set: SyntaxSet::load_defaults_nonewlines(),
            theme,
            options: PhantomData,
        }
    }
}

impl<'data, 'options, I: Iterator<Item = AnnotatedEvent<'data>>> Iterator
    for SyntectIter<'data, 'options, I>
{
    type Item = AnnotatedEvent<'data>;

    fn next(&mut self) -> Option<Self::Item> {
        let annotated_event = self.source.next()?;
        if let Event::CodeBlock(CodeBlockEvent {
            language: Some(ref language),
            ref code,
        }) = annotated_event.event
        {
            let language = language.as_str();
            let syntax = self
                .syntax_set
                .find_syntax_by_token(language)
                .unwrap_or_else(|| self.syntax_set.find_syntax_plain_text());
            let mut h = HighlightLines::new(syntax, &self.theme);
            let regions = h.highlight(code.as_str(), &self.syntax_set);
            return Some(AnnotatedEvent {
                event: Event::RawHtml(RawHtmlEvent {
                    html: format!(
                        "<pre><code>{}</code></pre>",
                        styled_line_to_highlighted_html(&regions[..], IncludeBackground::No)
                    )
                    .into(),
                }),
                location: annotated_event.location,
            });
        }
        Some(annotated_event)
    }
}
