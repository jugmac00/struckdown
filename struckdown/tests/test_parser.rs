use std::fs;

use struckdown::event::{AnnotatedEvent, DocumentStartEvent, Event};
use struckdown::parser::parse;
use struckdown::processors;

use itertools::Either;

fn apply_configured_extensions<'data, I: 'data + Iterator<Item = AnnotatedEvent<'data>>>(
    extensions: Vec<String>,
    iter: I,
) -> impl Iterator<Item = AnnotatedEvent<'data>> {
    let mut iter = Box::new(iter) as Box<dyn Iterator<Item = AnnotatedEvent<'data>>>;

    for extension in extensions {
        if extension == "AutoAnchors" {
            iter = Box::new(processors::AutoAnchors::new(iter));
        } else {
            panic!("unknown extension {}", extension);
        }
    }

    return iter;
}

fn apply_extensions<'data, I: 'data + Iterator<Item = AnnotatedEvent<'data>>>(
    iter: I,
) -> impl Iterator<Item = AnnotatedEvent<'data>> {
    let mut iter = iter.peekable();

    if let Some(document_start) = iter.peek() {
        match document_start.event {
            Event::DocumentStart(DocumentStartEvent {
                front_matter: Some(ref front_matter),
            }) => {
                if let Some(extensions) =
                    front_matter.get("extensions").and_then(|x| x.as_sequence())
                {
                    return Either::Left(apply_configured_extensions(
                        extensions
                            .iter()
                            .map(|x| x.as_str().unwrap().to_owned())
                            .collect(),
                        iter,
                    ));
                }
            }
            _ => {}
        }
    }

    Either::Right(iter)
}

#[test]
fn test_basics() {
    insta::glob!("inputs/*.md", |file| {
        let source = fs::read_to_string(file).unwrap();
        let events: Vec<_> = apply_extensions(parse(&source)).collect();
        insta::assert_yaml_snapshot!(events);
    });
}
