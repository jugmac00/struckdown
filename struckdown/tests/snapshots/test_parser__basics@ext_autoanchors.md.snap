---
source: struckdown/tests/test_parser.rs
expression: events
input_file: struckdown/tests/inputs/ext_autoanchors.md
---
- - type: document_start
    front_matter:
      extensions:
        - AutoAnchors
  - offset: 0
    len: 37
    line: 1
    column: 0
- - type: start_tag
    tag: heading1
    attrs:
      id: heading-1
  - offset: 0
    len: 12
    line: 1
    column: 0
- - type: start_tag
    tag: paragraph
  - offset: 13
    len: 7
    line: 3
    column: 0
- - type: text
    text: Text 1
  - offset: 13
    len: 6
    line: 3
    column: 0
- - type: end_tag
    tag: paragraph
  - offset: 13
    len: 7
    line: 3
    column: 0
- - type: start_tag
    tag: heading1
    attrs:
      id: manual-id
  - offset: 21
    len: 25
    line: 5
    column: 0
- - type: text
    text: Heading 2
  - offset: 23
    len: 9
    line: 5
    column: 2
- - type: end_tag
    tag: heading1
  - offset: 21
    len: 25
    line: 5
    column: 0
- - type: start_tag
    tag: paragraph
  - offset: 47
    len: 7
    line: 7
    column: 0
- - type: text
    text: Text 2
  - offset: 47
    len: 6
    line: 7
    column: 0
- - type: end_tag
    tag: paragraph
  - offset: 47
    len: 7
    line: 7
    column: 0
